src/python/pants/backend/project_info/rules/source_file_validator.py

```diff
@@ -0,0 +1,249 @@
+        matching.append(content_pattern_name)
+      else:
+        nonmatching.append(content_pattern_name)
+    return tuple(matching), tuple(nonmatching)
+
+  def get_applicable_content_pattern_names(self, path):
+    """Return the content patterns applicable to a given path.
+
+    Returns a tuple (applicable_content_pattern_names, content_encoding).
+
+    If path matches no path patterns, the returned content_encoding will be None (and
+    applicable_content_pattern_names will be empty).
+    """
+    encodings = set()
+    applicable_content_pattern_names = set()
+    for path_pattern_name, content_pattern_names in self._required_matches.items():
+      m = self._path_matchers[path_pattern_name]
+      if m.matches(path):
+        encodings.add(m.content_encoding)
+        applicable_content_pattern_names.update(content_pattern_names)
+    if len(encodings) > 1:
+      raise ValueError('Path matched patterns with multiple content encodings ({}): {}'.format(
+        ', '.join(sorted(encodings)), path
+      ))
+    content_encoding = next(iter(encodings)) if encodings else None
+    return applicable_content_pattern_names, content_encoding
+
+
+# TODO: Switch this to `lint` once we figure out a good way for v1 tasks and v2 rules
+# to share goal names.
+@console_rule('validate', [Console, RegexMatchResults, SourceFileValidation])
+def validate(console, regex_match_results, source_file_validation):
+  detail_level = source_file_validation.get_options().detail_level
+  regex_match_results = sorted(regex_match_results, key=lambda x: x.path)
+  num_matched_all = 0
+  num_nonmatched_some = 0
+  for rmr in regex_match_results:
+    if not rmr.matching and not rmr.nonmatching:
+      continue
+    if rmr.nonmatching:
+      icon = 'X'
+      num_nonmatched_some += 1
+    else:
+      icon = 'V'
+      num_matched_all += 1
+    matched_msg = ' Matched: {}'.format(','.join(rmr.matching)) if rmr.matching else ''
+    nonmatched_msg = (" Didn't match: {}".format(','.join(rmr.nonmatching))
+                      if rmr.nonmatching else '')
+    if (detail_level == DetailLevel.all or
+        (detail_level == DetailLevel.nonmatching and nonmatched_msg)):
+      console.print_stdout("{} {}:{}{}".format(icon, rmr.path, matched_msg, nonmatched_msg))
+
+  if detail_level != DetailLevel.none:
+    console.print_stdout('\n{} files matched all required patterns.'.format(num_matched_all))
+    console.print_stdout('{} files failed to match at least one required pattern.'.format(
+      num_nonmatched_some))
+
+  if num_nonmatched_some:
+    raise GracefulTerminationException('Files failed validation.')
+
+
+@rule(RegexMatchResults, [HydratedTargets, SourceFileValidation])
+def match_regexes(hydrated_targets, source_file_regexes):
+  multi_matcher = source_file_regexes.get_multi_matcher()
+  res = []
+  for tgt in hydrated_targets.dependencies:
+    if hasattr(tgt.adaptor, 'sources'):
+      files_content = yield Get(FilesContent,
```

Comment:
The effect of using `for-yield` like this is that some opportunity for parallelism is lost: see #7491.

Relatively little of this will be parallelizable since there are not external processes involved, and only a little bit of IO, but would recommend splitting out a rule that produces a `RegexMatchResult`, and then having one of these rules do:
```
regex_match_results = yield [Get(RegexMatchResult, HydratedTarget, ht) for ht in hydrated_targets]
```
instead.

Reply:
Great to know. Done.